File: frontend/src/api/knowledge.ts
import apiClient from './index'

export interface KnowledgeNodeItem {
  id: string
  name: string
  difficulty: number
  prerequisites: string[]
}

export interface KnowledgeCategory {
  category: string
  nodes: KnowledgeNodeItem[]
}

export interface KnowledgeTreeResponse {
  subject: string
  tree: KnowledgeCategory[]
}

export interface KnowledgeNodeDetail {
  id: string
  name: string
  subject: string
  category: string
  difficulty: number
  prerequisites: string[]
  concept_explanation: string | null
  common_pitfalls: string[]
  related_question_ids: string[]
}

export interface CrossLink {
  node_id: string
  node_name: string
  subject: string
  relation: string
}

export interface SubjectInfo {
  id: string
  name: string
}

export const knowledgeApi = {
  getTree(subject = 'ds'): Promise<KnowledgeTreeResponse> {
    return apiClient.get(`/knowledge/tree?subject=${subject}`)
  },

  getNode(nodeId: string): Promise<KnowledgeNodeDetail> {
    return apiClient.get(`/knowledge/nodes/${nodeId}`)
  },

  getNodeQuestions(nodeId: string): Promise<{ node_id: string; questions: Array<{ id: string; content: string; difficulty: number }> }> {
    return apiClient.get(`/knowledge/nodes/${nodeId}/questions`)
  },

  getCrossLinks(nodeId: string): Promise<CrossLink[]> {
    return apiClient.get(`/knowledge/nodes/${nodeId}/cross-links`)
  },

  getSubjects(): Promise<{ subjects: SubjectInfo[] }> {
    return apiClient.get('/knowledge/subjects')
  },
}

File: backend/app/schemas/knowledge.py
"""Knowledge schemas."""
from pydantic import BaseModel


class KnowledgeNodeItem(BaseModel):
    id: str
    name: str
    difficulty: int = 1
    prerequisites: list[str] = []

    class Config:
        from_attributes = True


class KnowledgeCategory(BaseModel):
    category: str
    nodes: list[KnowledgeNodeItem]


class KnowledgeTreeResponse(BaseModel):
    subject: str
    tree: list[KnowledgeCategory]


class CrossLinkItem(BaseModel):
    node_id: str
    node_name: str
    subject: str
    relation: str


class KnowledgeNodeDetail(BaseModel):
    id: str
    name: str
    subject: str
    category: str = ""
    difficulty: int = 1
    prerequisites: list[str] = []
    concept_explanation: str | None = None
    common_pitfalls: list[str] = []
    related_question_ids: list[str] = []

    class Config:
        from_attributes = True
